posthog/models/historical_version.py

```diff
@@ -0,0 +1,109 @@
+import datetime
+import json
+
+from django.db import models
+from django.utils import timezone
+
+from posthog.models import FeatureFlag
+from posthog.models.utils import UUIDModel
+
+
+class HistoricalVersionJSONEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, models.base.ModelState):
+            return None
+        if isinstance(obj, FeatureFlag):
+            return obj.__dict__
+        if isinstance(obj, datetime.datetime):
+            return obj.isoformat()
+
+        return json.JSONEncoder.default(self, obj)
+
+
+class HistoricalVersion(UUIDModel):
+    """
+     When calling delete lifecycle hooks in django rest framework
+    (see "Save and deletion hooks" in https://www.django-rest-framework.org/api-guide/generic-views/)
+
+    With Django Rest Framework you are not provided with the before and after states, so:
+     * storing the before and after of the model
+     * or computing and storing the change that occurs to the model
+    would require complication or extra DB reads
+
+    Django Rest Framework provides the state of the model _after_ the change.
+    So, this history log stores that state and uses it to compute what changed
+    when reading the history log
+    (see history_logging.py for how change is computed)
+    """
+
+    class Action(models.TextChoices):
+        Create = "create"
+        Update = "update"
+        Delete = "delete"
+
+    class Meta:
+        constraints = [
+            models.CheckConstraint(
+                check=models.Q(team_id__isnull=False) | models.Q(organization_id__isnull=False),
+                name="must_have_team_or_organization_id",
+            ),
+        ]
+        indexes = [
+            models.Index(fields=["item_id", "team_id", "name"]),
+        ]
+
+    # JSON of the historical item
+    state = models.JSONField(null=False, encoder=HistoricalVersionJSONEncoder)
```

Comment:
Will we be able to restore state from this?

Reply:
I think we should decide what we intend to be doing here, are we logging, or are we versioning. IMO we should keep versioning and "action" logging separate. We shouldn't be attempting to restore state from a log.

As far as I understand, the immediate requirement is a log.